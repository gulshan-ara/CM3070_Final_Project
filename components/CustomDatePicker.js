/** This component is used for rendering date picker.
 * Here the label prop handles the label of date picker
 * The pressable TextInput field handles the visibility of date picker.
 */

// import libraries and packages
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import React, { useState } from "react";
import DatePicker from "@react-native-community/datetimepicker";

const CustomDatePicker = ({ label, placeHolder, onDateChange }) => {
	// handles the visibility of date Picker
	const [showPicker, setShowPicker] = useState(false);
	// handle date state
	const [date, setDate] = useState(placeHolder);

	// function to handle date change
	const onChangeDate = ({ type }, pickedDate) => {
		// if date is set (Ok button is pressed)
		if (type === "set") {
			const currentDate = pickedDate;
			// converting date object into string value and storing in state
			setDate(currentDate.toDateString());
			// updating the given state value to use later
			onDateChange(currentDate.toDateString());
			// hiding date picker
			setShowPicker(false);
		} else {
			setShowPicker(!showPicker);
		}
	};

	// rendering the view
	return (
		<>
			{/* Render the date picker input box with label */}
			<View style={styles.container}>
				{/* customisable label for datePicker */}
				<Text style={styles.label}>{label} :</Text>
				{/* TouchableOpacity is used for handling the visibility of datePicker */}
				<TouchableOpacity
					style={styles.dateContainer}
					onPress={() => setShowPicker(true)}
				>
					{/* Text field to show the date */}
					<Text style={styles.dateText}>{date}</Text>
				</TouchableOpacity>
			</View>

			{/* Render the date picker on condition */}
			{showPicker && (
				<DatePicker
					mode="date" // mode defines the picker type(eg: date/time)
					display="calender" // picker view is defined by display prop
					value={new Date()} // value is required & recieves a date object to show on initial render
					onChange={onChangeDate} // handles the date change
				/>
			)}
		</>
	);
};

export default CustomDatePicker;

const styles = StyleSheet.create({
	container: {
		marginVertical: 5,
		marginHorizontal: 10,
		flexDirection: "row",
	},
	label: {
		fontWeight: "bold",
		fontSize: 18,
		margin: 5,
		paddingHorizontal: 5,
		flex: 1.5,
	},
	dateContainer: {
		backgroundColor: "rgba(255, 250, 205, 0.7)",
		flex: 2.5,
		marginHorizontal: 5,
		padding: 8,
		// alignItems: "center",
		borderBottomWidth: 1
	},
	dateText: {
		fontSize: 16,
		letterSpacing: 0.5,
	},
});
