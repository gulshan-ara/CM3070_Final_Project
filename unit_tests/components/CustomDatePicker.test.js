/** This files contains unit tests for the custom text input component */

// import libraries
import React from "react";
import { render, fireEvent } from "@testing-library/react-native";

// import component file
import CustomDatePicker from "../../components/CustomDatePicker";

/** unit test - 1
 *  The purpose of this unit test is to check whether the component is rendered properly or not.
 */

describe("CustomDatePicker", () => {
	it('render correctly datepicker component', () => {  
		const DatePickerComponent = render(<CustomDatePicker />).toJSON();
		expect(DatePickerComponent).toMatchSnapshot();
	});

	it("renders CustomDatePicker correctly", () => {
		// dummy values to pass as props
		const datePickerLabel = "datepicker label";
		// rendering the component
		const { getByText } = render(
			<CustomDatePicker label={datePickerLabel} />
		);
		// Assertions of successful render of the datepicker input label
		const labelTest = getByText(`${datePickerLabel} :`);
		expect(labelTest).toBeTruthy();
	});

});
