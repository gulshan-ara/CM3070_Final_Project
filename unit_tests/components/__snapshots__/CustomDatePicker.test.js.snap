// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDatePicker render correctly datepicker component 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "marginHorizontal": 10,
      "marginVertical": 5,
    }
  }
>
  <Text
    style={
      {
        "flex": 1.5,
        "fontSize": 18,
        "fontWeight": "bold",
        "margin": 5,
        "paddingHorizontal": 5,
      }
    }
  >
     :
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "rgba(255, 250, 205, 0.7)",
        "borderBottomWidth": 1,
        "flex": 2.5,
        "marginHorizontal": 5,
        "opacity": 1,
        "padding": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "letterSpacing": 0.5,
        }
      }
    />
  </View>
</View>
`;
